File: src/loggers/volume.rs
//! Logs volume change.
use super::{FifoRunner, Logger};
use std::env;

struct VolumeRunner;

impl FifoRunner for VolumeRunner {
    fn fmt_value(&mut self, string: &str) -> String {
        format!(
            "<ScrlU:Shift=volume 5%+><ScrlD:Shift=volume 5%-><ScrlU=volume 1%+><ScrlD=volume 1%-><BtnL=volume toggle> {}  </BtnL></ScrlD></ScrlU></ScrlD></ScrlU><Box:Left=#171717:2> </Box>",
            string
        )
    }
}

pub fn create_volume_logger() -> Logger {
    Logger::FifoLogger {
        default_value: "vol: ?".into(),
        fifopath: env::var("XDG_RUNTIME_DIR")
            .map(|dir| format!("{}/pipe/volume", dir))
            .unwrap(),
        runner: Box::new(VolumeRunner {}),
    }
}

File: logger/volume.go
package logger

import . "smolprog/utils"

type Volume struct{}

func (vol *Volume) Run(slot int, channel chan<- *Message) {
	default_value := vol.Fmt("  ?")
	msg := Message{Slot: slot, Value: default_value}

	for channel <- &msg; ; {
		if line, err := FirstLineOf(XDG_RUNTIME_DIR + "/pipe/volume"); err == nil {
			if len(line) > 0 {
				msg.Value = vol.Fmt(line)
			} else {
				msg.Value = default_value
			}
			channel <- &msg
		}
	}
}

func (_ *Volume) Fmt(s string) string {
	return SEP + "<ScrlU:Shift=volume 5%+>" +
		"<ScrlD:Shift=volume 5%->" +
		"<ScrlU=volume 1%+>" +
		"<ScrlD=volume 1%->" +
		"<BtnL=volume toggle>  " + s + "  </BtnL>" +
		"</ScrlD>" +
		"</ScrlU>" +
		"</ScrlD>" +
		"</ScrlU>"
}
